README.md
karthipad
=========
karthipad is an application to turn your touchpad/trackpad into a writing pad.
Once you run karthipad, you can write on touchpad as if you are writing on a paper.
The content what you write will be saved to a bitmap file (.bmp)

**Dependencies :** synclient

Initial Setup
=============

1. Type the following command in terminal
		locate synaptics
	You will find a file like "50-synaptics.conf"
	 either in /etc/X11/xorg.conf.d/ or /usr/share/X11/xorg.conf.d/

2. Open the file using sudo
	    example: >sudo gedit /etc/X11/xorg.conf.d/50-synaptics.conf

3. Add the following line (with exact spaces- don't use tab) and save the file
	        Option         "SHMConfig" "on"

4. Restart the system.

5. Type the follwing command in terminal
	    synclient -m 100
	 Now, when you move your finger over touchpad, you can see the values printed in the terminal.
	 Press Ctrl+C to stop.

Compilation
===========

Extract the source files.
Compile using the following command

	make

Recompile:
		make clean
		make

Running Binary:

	./karthipad

Installation
============
Install using following command

	make
	sudo make install

Un-install using following command

		sudo make uninstall

Re-install using following command

		sudo make reinstall
Usage
=====
Following commands are supported.

		karthipad filename
		#(default size = 1000x1000)
		karthipad filename 980
		#(size = 980x980)
		karthipad filename 999 500
		#(size = 999x500)
		karthipad -h
		#(displays help information)
		karthipad -v
		#(displays version information)
To quit the command line application, type q.

Support Information
===================

Modules used:

- libbmp http://code.google.com/p/libbmp/
- synclient http://en.gentoo-wiki.com/wiki/Synaptics_Touchpad

About Author
============
Karthikeyan.N
- Email: karthi.amrita.cbe@gmail.com
- Website:techbird.wordpress.com
- Twitter: www.twitter.com/lxkarthi


Changes.txt
karthipad 1.0.0
1. Created python based GUI version
2. Save image as jpg file
3. drawing in canvas and image file are seperate(not related).

karthipad 0.2.0
1. Automatically detect resolution of touchpad (approximate detection only).
2. linear interpolation for continous movement of finger (two point form line, width pixel width 3x3).
3. linear interpolation with two implementations, (x versus y) and (y versus x)
4. drawing pixel is put as seperate inline function.

karthipad 0.1.0
1. track the position of finger and draw a point.
2. command line input only.
3. disable mouse movement while drawing.



wpad.c
/**************** AUTHOR ****************/
/* Karthikeyan.N			*/
/* email:karthi.amrita.cbe@gmail.com 	*/
/****************************************/
/*
 Description:
 Draws a image file from synclient output until it sees a letter 'q'
 synclient -m 10 gives the touchpad status every 10 ms.
*/
#include<stdio.h>
#include<string.h>
#include "bmpfile.c"
int contains(char* str, char val);
inline bool draw_pixel(bmpfile_t *bmp, uint32_t x, uint32_t y, rgb_pixel_t pixel);
int main(int argc, char **argv)
{
  char* b;
  char a[200];
  float time;
  int x, y, z, f, w, l, r, u, d, m, multi;
  int prev_x,prev_y,prev_z,prev_f,x1,x2,y1,y2,i,j;
  bmpfile_t *bmp;
  rgb_pixel_t pixel = {128, 64, 0, 0};

  if( argc == 5 )
	printf("Output File: %s\n", argv[1]);
  else if( argc < 5 )
  {
	printf("Usage: %s filename width height depth.\n", argv[0]);
	return 1;
  }
  else
	printf("Too many arguments supplied.\n");

  if ((bmp = bmp_create(atoi(argv[2]), atoi(argv[3]), atoi(argv[4]))) == NULL) {
	printf("Invalid depth value: %s.\n", argv[4]);
	return 1;
  }

  prev_x=-1;  prev_y=-1;  prev_z=-1;  prev_f=0;
  f=0;
  while(1)
  {
	b=gets(a);
	if(contains(a,'q'))
		break;
	if(contains(a,'t')==0)
	{
		//time     x    y    z f  w  l r u d m     multi  gl gm gr gdx gdy
		//0.000   539  532   0 0  0  0 0 0 0 0  00000000
		sscanf(a,"%f %d %d %d %d %d %d %d %d %d %d %d", &time, &x, &y, &z, &f, &w, &l, &r, &u, &d, &m, &multi);
/*TODO Pixel value change according to pressure*/
/*		pixel.red=z;
		pixel.green++;
		pixel.blue++;
*/
/*Draw line between consecutive points on moving (use 'f' value to check finger is continuously moving or not)*/
/*TODO curve (second order interpolation) So, smoother drawing*/
		//Setting Pixels around the touching point
		if(f==1)
		{
			draw_pixel(bmp, x, y, pixel);
			//Draw line if you move finger without removing from surface.
			if(prev_f==1)
			{
				if( abs(prev_x-x) > abs(prev_y-y) )
				{
					if(x>prev_x) { x2=x; x1=prev_x; y2=y; y1=prev_y;}
					else {x2=prev_x; x1=x; y2=prev_y; y1=y; }
					// Line drawing between two points (x versus y)
					for(i=x1;i<x2;i++)
					{
					j= y1 + (((y2-y1)*(i-x1))/(x2-x1));
					draw_pixel(bmp, i, j, pixel);
					}
				}
				else
				{
					if(y>prev_y) { x2=x; x1=prev_x; y2=y; y1=prev_y;}
					else {x2=prev_x; x1=x; y2=prev_y; y1=y; }
					// Line drawing between two points (y versus x)
					for(j=y1;j<y2;j++)
					{
					i= x1 + (((x2-x1)*(j-y1))/(y2-y1));
					draw_pixel(bmp, i, j, pixel);
					}
				}
			}
		}
		prev_x=x;  prev_y=y;  prev_z=z;  prev_f=f;
//Debug		printf("%d. %f %d %d %d %d %d %d %d %d %d %d %d\n",i++,time,x, y, z, f, w, l, r, u, d, m, multi);
	}
  }
  bmp_save(bmp, argv[1]);
  bmp_destroy(bmp);
  printf("Drawing Finished\n");
  return 0;
}

int contains(char* str, char val)
{
  int j,len;
  len=strlen(str);
  for(j=0;j<len;j++)
  {
	if(str[j]==val)
	return 1;
  }
  return 0;
}

inline bool draw_pixel(bmpfile_t *bmp, uint32_t x, uint32_t y, rgb_pixel_t pixel)
{
	/*For more thickness, adjacent pixels are also set.*/
	bmp_set_pixel(bmp, x+1, y, pixel);
	bmp_set_pixel(bmp, x, y+1, pixel);
	bmp_set_pixel(bmp, x-1, y, pixel);
	bmp_set_pixel(bmp, x, y-1, pixel);
	bmp_set_pixel(bmp, x-1, y-1, pixel);
	bmp_set_pixel(bmp, x+1, y+1, pixel);
	/*For thinner lines, remove above lines in this function*/	
	return bmp_set_pixel(bmp, x, y, pixel);
}


draw.py
from Tkinter import *
class PaintBox(Frame):
	"""Demonstrate drawing on a Canvas"""

	def __init__( self ):
		"""Create Canvas and bind paint method to mouse dragging"""
		Frame.__init__( self )
		self.pack( expand = YES, fill = BOTH )
		self.master.title( "A simple paint program" )
		self.master.geometry( "300x300" )

		self.message = Label( self,
		text = "Click and Drag the mouse to draw" )
		self.message.pack( side = BOTTOM )
	# create Canvas component
		self.myCanvas = Canvas(self, bg="white", width=300, height= 300) #Canvas( self )
		self.myCanvas.pack( expand = YES, fill = BOTH )
	# bind mouse dragging event to Canvas
		self.myCanvas.bind( "<B1-Motion>", self.paint )

	def paint( self, event ):
		"""Create an oval of radius 4 around the mouse position"""
		x1, y1 = ( event.x - 4 ), ( event.y - 4 )
		x2, y2 = ( event.x + 4 ), ( event.y + 4 )
		self.myCanvas.create_oval( x1, y1, x2, y2, fill = "black" )

def main():
	PaintBox().mainloop()

if __name__ == "__main__":
	main()


karthipadgui.py
#!/usr/bin/python
import sys
import Tkinter as tk
import Image, ImageDraw
from threading import Thread
import time

width = 980/2
height = 720/2
#print 'Number of arguments:', len(sys.argv), 'arguments.'
print 'Argument List:', str(sys.argv)
args= sys.argv
width = int(args[2])
height= int(args[3])
filename= args[1]
center = height//2
white = (255, 255, 255)
green = (0,128,0)

class App():
    def __init__(self):
	self.root = tk.Tk()
	self.label = tk.Label(text="")
	self.label.pack()
	# Tkinter create a canvas to draw on
	self.cv = tk.Canvas(self.root, width=width/2, height=height/2, bg='white')
	self.cv.pack()

	# PIL create an empty image and draw object to draw on
	# memory only, not visible
	self.image1 = Image.new("RGB", (width, height), white)
	self.draw = ImageDraw.Draw(self.image1)

	# do the Tkinter canvas drawings (visible)
	#self.cv.create_line([0, center, width, center], fill='green')

	# do the PIL image/draw (in memory) drawings
	#self.draw.line([0, center, width, center], green)
	#self.draw.ellipse([50, 45, 55, 60], fill=green)

	#sync variable for thread loop
	self.goon=1
	try:
		Thread(target=self.listen ).start()
	except Exception, errtxt:
		print errtxt
	self.update_clock()
	self.root.protocol("WM_DELETE_WINDOW", self.handler)
	self.root.mainloop()

    def update_clock(self):
	now = time.strftime("%H:%M:%S")
	self.label.configure(text=now)
	self.root.after(1000, self.update_clock)

    def listen(self):
	#self.cv.create_line([50, center+50, width, center], fill='green')
	s=raw_input()
	while( ("q" not in s) and self.goon==1):
		if("time" not in s):
			coords=s.split()
			#   time     x    y   z f  w  l r u d m     multi  gl gm gr gdx gdy
			if(coords[4]=='1'):
				x=int(coords[1])
				y=int(coords[2])
				z=int(coords[3])
				#self.cv.create_line([x, y, x+1, y+1], fill='green')
				#self.cv.create_rectangle([x, y, x+z/4, y+z/4], outline='green', fill='green')
				self.cv.create_oval([x/2, y/2, x/2+z/8, y/2+z/8], outline='green', fill='green')
				self.draw.ellipse([x-z/8, y-z/8, x+z/8, y+z/8], fill=green)
				#print coords
		s=raw_input()

    def handler(self):
	self.goon=0
	print "Killed soon"
	# PIL image can be saved as .png .jpg .gif or .bmp file (among others)
	#filename = "my_drawing.jpg"
	self.image1.save(filename)
	self.root.quit()

app=App()
#app.mainloop()

